src/cli/setup_mode.rs

```diff
@@ -1,112 +1,92 @@
 use anyhow::Result;
-use clap::{builder::PossibleValuesParser, Arg, ArgAction, Command};
+use clap::{builder::PossibleValuesParser, Parser};
 
 use crate::{
     cli::{
         common,
         self_update::{self, InstallOpts},
     },
-    currentprocess::{argsource::ArgSource, filesource::StdoutSource},
+    currentprocess::filesource::StdoutSource,
     dist::dist::Profile,
     process,
     toolchain::names::{maybe_official_toolchainame_parser, MaybeOfficialToolchainName},
     utils::utils,
 };
 
-#[cfg_attr(feature = "otel", tracing::instrument)]
-pub fn main() -> Result<utils::ExitCode> {
-    let args: Vec<_> = process().args().collect();
-    let arg1 = args.get(1).map(|a| &**a);
+/// The installer for rustup
+#[derive(Parser)]
+#[command(
+    name = "rustup-init",
+    bin_name = "rustup-init[EXE]",
+    version = common::version(),
+    before_help = format!("rustup-init {}", common::version()),
+)]
+struct RustupInit {
+    /// Enable verbose output
+    #[arg(short, long)]
+    verbose: bool,
 
-    // Secret command used during self-update. Not for users.
-    if arg1 == Some("--self-replace") {
-        return self_update::self_replace();
-    }
+    /// Disable progress output
+    #[arg(short, long)]
+    quiet: bool,
 
-    // Internal testament dump used during CI.  Not for users.
-    if arg1 == Some("--dump-testament") {
-        common::dump_testament()?;
-        return Ok(utils::ExitCode(0));
-    }
+    /// Disable confirmation prompt
+    #[arg(short = 'y')]
+    no_prompt: bool,
+
+    /// Choose a default host triple
+    #[arg(long)]
+    default_host: Option<String>,
+
+    /// Choose a default toolchain to install. Use 'none' to not install any toolchains at all
+    #[arg(long, value_parser = maybe_official_toolchainame_parser)]
+    default_toolchain: Option<MaybeOfficialToolchainName>,
+
+    #[arg(long, value_parser = PossibleValuesParser::new(Profile::names()), default_value = Profile::default_name())]
+    profile: String,
+
+    /// Component name to also install
+    #[arg(short, long, value_delimiter = ',', num_args = 1..)]
+    components: Option<Vec<String>>,
```

Comment:
You've probably guessed it: this happens for a reason.

Here I'm sticking to the type in the original code because this allows us to generate `-t, --targets <TARGETS>...` in the help string.

The initial version of this PR used `Vec<String>` and instead it generated something like `-t, --targets <TARGETS>`.

Reply:
So this is just to keep the `...` in the CLI docs around? IMO it would be better to just lose the `...`.

Reply:
@djc I believe there's also a corner case where someone could run `rustup-init targets=`. This is forbidden by the original but would be allowed by `Vec<String>`. Do we want that though?

Reply:
@djc Oh, I think I've found a way to do that properly: `Vec<String>` can be used with `num_args = 1..` and we can get the best of both worlds. I'll do that instead.